pkg/controller/nodeipam/ipam/multicidrset/multi_cidr_set.go

```diff
@@ -0,0 +1,365 @@
+		nodeMask:         net.CIDRMask(subNetMaskSize, bits),
+		clusterMaskSize:  clusterMaskSize,
+		MaxCIDRs:         1 << uint32(subNetMaskSize-clusterMaskSize),
+		NodeMaskSize:     subNetMaskSize,
+		Label:            cidrConfig.String(),
+		AllocatedCIDRMap: make(map[string]bool, 0),
+	}, nil
+}
+
+func (s *MultiCIDRSet) indexToCIDRBlock(index int) (*net.IPNet, error) {
+	var ip []byte
+	switch /*v4 or v6*/ {
+	case s.ClusterCIDR.IP.To4() != nil:
+		{
+			j := uint32(index) << uint32(32-s.NodeMaskSize)
+			ipInt := (binary.BigEndian.Uint32(s.ClusterCIDR.IP)) | j
+			ip = make([]byte, net.IPv4len)
+			binary.BigEndian.PutUint32(ip, ipInt)
+		}
+	case s.ClusterCIDR.IP.To16() != nil:
+		{
+			// leftClusterIP      |     rightClusterIP
+			// 2001:0DB8:1234:0000:0000:0000:0000:0000
+			const v6NBits = 128
+			const halfV6NBits = v6NBits / 2
+			leftClusterIP := binary.BigEndian.Uint64(s.ClusterCIDR.IP[:halfIPv6Len])
+			rightClusterIP := binary.BigEndian.Uint64(s.ClusterCIDR.IP[halfIPv6Len:])
+
+			ip = make([]byte, net.IPv6len)
+
+			if s.NodeMaskSize <= halfV6NBits {
+				// We only care about left side IP.
+				leftClusterIP |= uint64(index) << uint(halfV6NBits-s.NodeMaskSize)
+			} else {
+				if s.clusterMaskSize < halfV6NBits {
+					// see how many bits are needed to reach the left side.
+					btl := uint(s.NodeMaskSize - halfV6NBits)
+					indexMaxBit := uint(64 - bits.LeadingZeros64(uint64(index)))
+					if indexMaxBit > btl {
+						leftClusterIP |= uint64(index) >> btl
+					}
+				}
+				// the right side will be calculated the same way either the
+				// subNetMaskSize affects both left and right sides.
+				rightClusterIP |= uint64(index) << uint(v6NBits-s.NodeMaskSize)
+			}
+			binary.BigEndian.PutUint64(ip[:halfIPv6Len], leftClusterIP)
+			binary.BigEndian.PutUint64(ip[halfIPv6Len:], rightClusterIP)
+		}
+	default:
+		return nil, fmt.Errorf("invalid IP: %s", s.ClusterCIDR.IP)
+	}
+	return &net.IPNet{
+		IP:   ip,
+		Mask: s.nodeMask,
+	}, nil
+}
+
+// NextCandidate returns the next candidate and the last evaluated index
+// for the current cidrSet. Returns nil if the candidate is already allocated.
+func (s *MultiCIDRSet) NextCandidate() (*net.IPNet, int, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.allocatedCIDRs == s.MaxCIDRs {
+		return nil, 0, &CIDRRangeNoCIDRsRemainingErr{
+			CIDR: s.Label,
+		}
+	}
+
+	candidate := s.nextCandidate
+	for i := 0; i < s.MaxCIDRs; i++ {
+		nextCandidateCIDR, err := s.indexToCIDRBlock(candidate)
+		if err != nil {
+			return nil, i, err
+		}
+		// Check if the nextCandidate is not already allocated.
+		if _, ok := s.AllocatedCIDRMap[nextCandidateCIDR.String()]; !ok {
+			s.nextCandidate = (candidate + 1) % s.MaxCIDRs
+			return nextCandidateCIDR, i, nil
+		}
+		candidate = (candidate + 1) % s.MaxCIDRs
+	}
+
+	return nil, s.MaxCIDRs, &CIDRRangeNoCIDRsRemainingErr{
+		CIDR: s.Label,
+	}
+}
+
+func (s *MultiCIDRSet) getBeginningAndEndIndices(cidr *net.IPNet) (int, int, error) {
```

Comment:
you should docs on the returned beginning, end indicies.

is it inclusive or exclusive of the end index?
e.g. is it "[0, 10) or [0, 10]"

Reply:
Added a comment